lib/install/management/configure.go

```diff
@@ -98,49 +102,146 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
 
 	snapshotName := fmt.Sprintf("%s %s", ConfigurePrefix, conf.Version.BuildNumber)
 	snapshotName = strings.TrimSpace(snapshotName)
-
 	// check for old snapshot
-	// #nosec: Errors unhandled.
-	oldSnapshot, _ := d.appliance.GetCurrentSnapshotTree(d.op)
+	oldSnapshot, err := d.appliance.GetCurrentSnapshotTree(d.op)
+	if err != nil {
+		// log the error but continue
+		d.op.Debugf("Error checking appliance snapshot tree during %s: %s", d.Action.String(), err.Error())
+	}
 
 	newSnapshotRef, err := d.createSnapshot(snapshotName, "configure snapshot")
 	if err != nil {
-		d.deleteUpgradeImages(ds, settings)
+		d.deleteISOs(ds, settings)
 		return err
 	}
 
-	err = d.update(conf, settings, isConfigureOp)
+	err = d.update(conf, settings)
+	if err != nil {
+		// Roll back
+		d.op.Errorf("Failed to %s: %s", d.Action.String(), err)
+		d.op.Infof("Rolling back %s", d.Action.String())
+
+		if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
+			d.op.Errorf("Failed to revert appliance to snapshot: %s", rerr)
+			return err
+		}
+		d.op.Infof("Appliance is rolled back to previous version")
+		d.deleteISOs(ds, settings)
+		d.deleteSnapshot(newSnapshotRef, snapshotName, conf.Name)
+
+		return err
+	}
+
+	// if we are upgrading evaluate need for inventory upgrade
+	if d.Action == UpgradeAction && d.session.IsVC() {
+		err := d.updateInventory(conf.Name)
+		if err != nil {
+			// if we returned an error then no Inventory work was
+			// started.  If we failed during update we warn the user.
+			d.op.Errorf("An error prevented inventory changes: %s", err)
+		}
+	}
 
 	// If successful try to grant permissions to the ops-user
-	if err == nil && conf.ShouldGrantPerms() {
-		err = opsuser.GrantOpsUserPerms(d.op, d.session, conf)
+	if conf.ShouldGrantPerms() {
+		// TODO: In a DRS environment that's not been upgraded to 1.4
+		// this will attempt to apply permissions to the VmFolder. Does
+		// that work & do we really want that?
```

Comment:
That _seems_ desirable, but we should probably have a robot test to make sure it works. (I'm not sure we have CI tests for the permissions functionality, but now that CI is on VC, it should be possible to add one.)

cc @lcastellano

Reply:
I _assumed_ the `VmFolder` -- which is datacenter wide -- was to broad and we'd not want to do it there.  But I'm not fluent on the permission code either, so I might be all wrong...